DuckDuckGo/BookmarksManager.swift

```diff
@@ -165,22 +165,25 @@ class BookmarksManager {
     
     func delete(_ bookmarkItem: BookmarkItem, completion: BookmarkItemDeletedBackgroundThreadCompletion? = nil) {
         var reloadWidget = false
+        var bookmarkURL: URL?
         if let bookmark = bookmarkItem as? Bookmark {
-            removeFavicon(forBookmark: bookmark)
             if bookmark.isFavorite {
                 reloadWidget = true
             }
+            bookmarkURL = bookmark.url
         }
-        coreDataStorage.delete(bookmarkItem.objectID) { success, error in
+        coreDataStorage.delete(bookmarkItem.objectID) { [weak self] success, error in
+            guard let self = self else { return }
             if reloadWidget {
                 self.reloadWidgets()
             }
+            self.removeFavicon(forBookmarkURL: bookmarkURL)
             completion?(success, error)
         }
     }
     
-    private func removeFavicon(forBookmark bookmark: Bookmark?) {
-        guard let domain = bookmark?.url?.host else { return }
+    private func removeFavicon(forBookmarkURL url: URL?) {
+        guard let domain = url?.host else { return }
         
         coreDataStorage.allBookmarksAndFavoritesFlat { bookmarks in
```

Comment:
Is it ever possible coreDataStorage.allBookmarksAndFavoritesFlat is fetched before the deletion is properly committed to core data? (I don't remember how core data handles that kind of thing, I think the answers no but want to double check)

Reply:
we call the completion block after a successful save, so the core data has already processed the changes and only then we call the method removeFavicon which refetches the model state, we should be safe here!